Python/dtoa.c

```diff
@@ -2563,6 +2563,13 @@ _Py_dg_dtoa(double dd, int mode, int ndigits,
                         }
                     ++*s++;
                 }
+                /* This branch was missing from the original dtoa.c, leading
+                   to surplus trailing zeros in some cases.
+                   See bugs.python.org/issue40780. */
+                else {
+                    while(*--s == '0');
```

Comment:
This block was copied, style-and-all, from the code above (line 2521), which does similar fix-up for the "fast path" that tries to get by with all-floating-point arithmetic. Correctness relies on `s` not consisting entirely of zero digits at this point. I _believe_ that that can't happen, but am still working through the logic.

Reply:
Added a note on why this is safe, along with an extra assert.

In general, if `d` is the double value we're looking for, then at the point of entry into this `for` loop, `d` is strictly positive (signs and zeros have already been taken care of earlier) and we have `10**(k-1) < d < 10**(k+1)`  (`k` is calculated earlier as `floor(log10(d))`, but the calculation uses an approximation, and `k` *could* a priori possibly be one too large.)

But if we're in this branch, then we already know that `d` is integral and that `k <= 14`. Scanning back up the code, if `k < 22` (`Ten_pmax`) then we've already done a direct comparison with the corresponding exactly representable power of 10 to fix `k` if necessary. So in fact, if we get here, `k` _cannot_ be one too large, so we know `10**k <= d`. On the first iteration of the for loop, `dval(&u)` is equal to `d`, `ds` is equal to `10**k`, and so `dval(&u) / ds` is at least `1` and the first digit recorded in `s` can't be a `0`.

Reply:
Not sure I'm comfortable with this twisty, fragile, non-local logic. I may add back an explicit check so that the zero-stripping can't go past the first digit, no matter what it is.

Reply:
Yes, that's probably more future-proof. It's belt and suspenders thing.

Reply:
Rewritten.